File: kimdokyun/C++/boj_1389.cpp
#include<stdio.h>

int graph[101][101];

int main() {
    int N,M;
    scanf("%d %d",&N,&M);
    for(int i=1;i<=N;i++) for(int j=1;j<=N;j++) {
        graph[i][j] = i==j?0:9999;
    }
    for(int i=0;i<M;i++) {
        int a,b;
        scanf("%d %d",&a,&b);
        graph[a][b] = graph[b][a] = 1;
    }
    for(int k=1;k<=N;k++) {
        for(int i=1;i<=N;i++) {
            for(int j=1;j<=N;j++) {
                int pass_k = graph[i][k] + graph[k][j];
                graph[i][j] = graph[i][j] > pass_k? pass_k:graph[i][j];
            }
        }
    }
    int result = 0, min = 9999;
    for(int i=1;i<=N;i++) {
        int sum = 0;
        for(int j=1;j<=N;j++) sum+=graph[i][j];
        if(min > sum) {
            result = i;
            min = sum;
        }
    }
    printf("%d",result);
}

File: kimdokyun/Python/boj_1389.py
graph = [[9999] * 101 for i in range(101)]

N,M = map(int, input().split())
for i in range(M):
    a,b = map(int, input().split())
    graph[a][b] = 1
    graph[b][a] = 1
for k in range(1, N+1):
    for i in range(1, N+1):
        for j in range(1, N+1):
            graph[i][j] = min(graph[i][j], graph[i][k]+graph[k][j])

print( min( [ (sum(graph[i]), i) for i in range(1, N+1) ] )[1] )